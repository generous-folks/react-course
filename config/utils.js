const getExerciseFromProcess = () => {
  const exNumber = process.argv[2];
  if(!process.env.APP_EXERCISE || !exNumber) {
    return null;
  }

  return exNumber;
}

module.exports = getExerciseFromProcess;
