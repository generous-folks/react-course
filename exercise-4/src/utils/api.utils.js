const endpoint = '/fixtures/articles.json';

export const getArticles = () =>
  fetch(endpoint)
    .then(res => res.json())
    .catch(err => console.log(err));
