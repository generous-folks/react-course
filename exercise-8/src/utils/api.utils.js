const endpoint = '/fixtures/articles.json';

export const getArticles = () =>
  fetch(endpoint)
    .then(res => res.json())
    .catch(err => console.log(err));

export const signIn = () => Promise.resolve({ id: 'xyz', mail: 'foo@bar.com', name: 'Foo Bar' });
export const signOut = args => Promise.resolve(args);
