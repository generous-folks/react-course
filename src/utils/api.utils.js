const endpoint = 'https://us-central1-gofire-1280e.cloudfunctions.net/ArticleAPI';

export const getArticles = () =>
  fetch(endpoint)
    .then(res => res.json())
    .catch(err => console.log(err));

export const setArticle = article =>
  fetch(endpoint, { method: 'POST', body: article })
    .then(res => res.json())
    .catch(err => console.log(err));

export const updateArticle = article =>
  fetch(endpoint, { method: 'PUT', body: article })
    .then(res => res.json())
    .catch(err => console.log(err));

export const deleteArticle = article =>
  fetch(endpoint, { method: 'DELETE', body: article })
    .then(res => res.json())
    .catch(err => console.log(err));
