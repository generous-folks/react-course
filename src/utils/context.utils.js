export const dispatchThunk = dispatch => param => {
  if (typeof param === 'function') {
    return param(dispatch);
  }

  return dispatch(param);
}
