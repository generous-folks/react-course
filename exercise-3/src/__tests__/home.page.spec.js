import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';

import { shallow } from 'enzyme';

import { HomePage } from '../pages/home.page';

import MockedArticles from '../../../fixtures/articles.json';
import * as ApiUtils from '../utils/api.utils';

let wrapper;
const emptyArray = [];

ApiUtils.getArticles = jest.fn().mockResolvedValue(MockedArticles);

jest.mock('react', () => global.mockReactWithHooks({ effect: true, state: true }));

const getWrapper = () => shallow(<HomePage />);

beforeEach(() => {
  jest.clearAllMocks();
  useEffect.mockClear();
  useState.mockClear();
  wrapper = getWrapper();
});

describe('<HomePage />', () => {
  describe('Snapshot', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Markup checks', () => {
    it('should contain the correct markup', () => {
      expect(wrapper.find(`[data-testid='app']`).exists()).toBeTruthy();
      expect(wrapper.find(`[data-testid='app-title']`).text()).toBe('Home Page');
      expect(wrapper.find(`[data-testid='articles-container']`).exists()).toBeTruthy();
      expect(wrapper.find(`[data-testid='articles-title']`).text()).toBe('Articles');
      expect(wrapper.find(`[data-testid='articles-list']`).exists()).toBeTruthy();
      expect(
        wrapper
          .find(Link)
          .first()
          .prop('to'),
      ).toBe('/about');
      expect(
        wrapper
          .find(Link)
          .last()
          .prop('to'),
      ).toBe('/contact');
    });
  });

  describe('State checks', () => {
    // Don't test Jest mocks or React, it is useless
    // In the following example we test that our react mock is working
    // const [initialsArticles] = useState.mock.results[0].value;
    // expect(initialsArticles).toEqual(emptyArray);

    it('should call useState with an empty array', () => {
      expect(useState).toHaveBeenCalledWith(emptyArray);
    });
  });

  describe('Effects checks', () => {
    it('should call useEffect with a function and an empty array', () => {
      expect(useEffect).toHaveBeenCalledWith(expect.any(Function), [emptyArray]);
    });

    it('should call getArticles once', () => {
      expect(ApiUtils.getArticles).toHaveBeenCalledTimes(1);
    });

    it('should call getArticles once even when re-rendered', () => {
      wrapper.update();
      expect(ApiUtils.getArticles).toHaveBeenCalledTimes(1);
    });

    it('should call setArticles once with MockedArticles', () => {
      const [, setArticles] = useState.mock.results[0].value;
      expect(setArticles).toHaveBeenCalledWith(MockedArticles);
      expect(setArticles).toHaveBeenCalledTimes(1);
    });
  });
});
