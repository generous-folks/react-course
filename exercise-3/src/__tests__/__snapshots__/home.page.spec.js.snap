// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> Snapshot should render correctly 1`] = `
<div
  data-testid="app"
>
  <h2
    data-testid="app-title"
  >
    Home Page
  </h2>
  <Link
    to="/about"
  >
    About Page
  </Link>
  <Link
    to="/contact"
  >
    Contact Page
  </Link>
  <div
    data-testid="articles-container"
  >
    <h4
      data-testid="articles-title"
    >
      Articles
    </h4>
    <ul
      data-testid="articles-list"
    />
  </div>
</div>
`;
